### PR TITLE
Add the 'eventually' convenience method

### PR DESCRIPTION
**Pull request by mweiden**
Eventually is a convenience method for testing against concurrent services. It retries a a function with assertions on an interval until those assertions pass or a timeout is reached.

### CODE REVIEW

File: tests/__init__.py [hunk 1]
```diff
@@ -1,5 +1,35 @@
...
+
+
+def eventually(timeout: float, interval: float):
+    """
+    eventually is designed to run a test multiple times until a the assertion is satisfied or a timeout is reached.
+    :param timeout: time until the test fails
+    :param interval: time between attempts of the test
+    :return: the result of the function or a raised assertion error
+    """
+    def decorate(func):
```
Comment (ttung) [inline tests/__init__.py]: Switch to:

```
@functools.wraps(func)
def call(*args, **kwargs):
```

File: tests/__init__.py [hunk 2]
```diff
@@ -1,5 +1,35 @@
...
+        def call(*args, **kwargs):
+            timeout_time = time.time() + timeout
+            result = None
```
Comment (ttung) [inline tests/__init__.py]: consider:

```
while True:
    try:
        return func(*args, **kwargs)
    except AssertionError:
        if time.time() >= timeout_time:
            raise
        time.sleep(interval)
```

**Comment by mweiden** [comment]
@ttung changes made 👍

File: tests/__init__.py [hunk 3]
```diff
@@ -1,5 +1,29 @@
...
+import functools
 
 
 def get_version():
     return datetime.datetime.utcnow().strftime("%Y-%m-%dT%H%M%S.%fZ")
+
+
+def eventually(timeout: float, interval: float):
+    """
+    eventually is designed to run a test multiple times until a the assertion is satisfied or a timeout is reached.
```
Comment (hannes-ucsc) [inline tests/__init__.py]: s/a the/the

Comment (hannes-ucsc) [reply tests/__init__.py]: Or actually `until all assertions are satisfied`

**Comment by kislyuk** [comment]
Consider making the caught exceptions configurable (by default, `set([AssertionError])`).

File: tests/__init__.py (hunk 1 above)
Comment (mweiden) [reply tests/__init__.py]: This change has been made.

File: tests/__init__.py (hunk 3 above)
Comment (mweiden) [reply tests/__init__.py]: Good catch. Made this change too.

**Comment by mweiden** [comment]
@kislyuk Good suggestion, I've made this change too.

File: tests/__init__.py [hunk 4]
```diff
@@ -1,5 +1,33 @@
 import datetime
+import time
```
Comment (ttung) [inline tests/__init__.py]: sort, pls.

File: tests/__init__.py [hunk 5]
```diff
@@ -1,5 +1,33 @@
...
+
+def eventually(timeout: float, interval: float, errors: set={AssertionError}):
+    """
+    @eventually runs a test until all assertions are satisfied or a timeout is reached.
+    :param timeout: time until the test fails
+    :param interval: time between attempts of the test
+    :param errors: the exceptions to catch and retry on
+    :return: the result of the function or a raised assertion error
+    """
+    def decorate(func):
```
Comment (ttung) [inline tests/__init__.py]: I think you can just remove L18, and `eventually(..)` can return `call`.

Comment (ttung) [reply tests/__init__.py]: nm, you're right.

**Pull request merged**